virtualization-and-containers/material/exercise_vm_containers_text.md

```diff
@@ -0,0 +1,175 @@
+- [Ubuntu Server download page](https://ubuntu.com/download/server)
+- [VirtualBox Manual](https://www.virtualbox.org/manual/UserManual.html)
+- If you want to set up SSH forwarding to your VM, you can find instructions, for example, [on the codebot homepage](https://codebots.com/docs/ubuntu-18-04-virtual-machine-setup). The instructions are for Ubuntu 18.04, but it also works for Ubuntu 20.04. Note that most of the steps are preconfigured in current VirtualBox releases, but it does not hurt to check whether the settings are set properly.
+
+
+## Virtual Machines Using Vagrant
+
+In the previous section we have set up a VM manually. This was quite tedious. Therefore, we want to automatize this process now by using [Vagrant](https://www.vagrantup.com/) and setting up our own Vagrant box. We will provision a box based on VirtualBox by using shell scripts and the included provisioning tools of Vagrant.
+
+### Tasks (Vagrant)
+
+- Fork the repository [TODO: "Exercise Virtual Machines"](https://gitlab-sim.informatik.uni-stuttgart.de/sse-test-group/exercise-virtual-machines) from GitLab and create a branch to work on the fork. It will contain a `README.md` and a file called `bootstrap.sh`.
+- We want to start from scratch so initialize a new box using `vagrant init` in inside this repository and add the resulting `Vagrantfile` to Git. Then adapt your `Vagrantfile` to incorporate the following settings:
+  - Your virtual machine must be based on the [`ubuntu/focal64` image](https://www.vagrantup.com/docs/boxes#official-boxes).
+  - The name of your VM must be `USERNAME-ubuntu-server`.
+  - The [box version](https://www.vagrantup.com/docs/boxes/versioning) must be `0.1.0`.
+  - The VM must request [384 MB of main memory](https://www.vagrantup.com/docs/providers/virtualbox/configuration).
+  - Run your box with `vagrant up` and make sure that everything works out as expected (`vagrant ssh`). If everything is fine, you can leave the VM and stop it.
+- In the next step we want to provision your virtual machines. That means we want to install additional software, add extra files, and set some additional variables. We want to provision (`config.vm.provision`) the box in several steps.
+  - Create a new file with the name `testfile` and add it to the repository. The file must contain your GitLab username as text. Add the file to your box using the [file provisioner](https://www.vagrantup.com/docs/provisioning/file) and place it in the home directory of the `vagrant` user.
+  - We want an [additional shared folder](https://www.vagrantup.com/docs/synced-folders/basic_usage), besides `/vagrant`, in our virtual machine. The directory where the `Vagrantfile` resides, i.e. `.`, should be mounted as `/mnt/shared/` in your container.
+  - Extend the file `bootstrap.sh` for provisioning your box. It should contain commands that you would usually use on the command line like in the VirtualBox task. The script must do the follwing:
```

Comment:
I get an error
```
/tmp/vagrant-shell: line 1: !/usr/bin/env: No such file or directory
```

Reply:
I guess I deleted a `#` too much from the `bootstrap.sh` when cleaning it up some time.

Reply:
This is fixed now in the GitLab repository.